File: cxsever/www/python/lib/lineardoc/text_block.py
"""
TextBlock - A block of annotated inline text.
"""

import re

from . import utils as Utils
from .text_chunk import TextChunk


class TextBlock:
    """A block of annotated inline text."""

    def __init__(self, text_chunks, can_segment=True):
        """
        Initialize a TextBlock.

        Args:
            text_chunks: Annotated inline text
            can_segment: This is a block which can be segmented
        """
        self.text_chunks = text_chunks
        self.can_segment = can_segment
        self.offsets = []

        cursor = 0
        for text_chunk in self.text_chunks:
            self.offsets.append({"start": cursor, "length": len(text_chunk.text), "tags": text_chunk.tags})
            cursor += len(text_chunk.text)

    def get_tag_offsets(self):
        """
        Get the start and length of each non-common annotation.

        Returns:
            Array of offset dicts
        """
        common_tags = self.get_common_tags()
        result = []
        for i, offset in enumerate(self.offsets):
            text_chunk = self.text_chunks[i]
            if len(text_chunk.tags) > len(common_tags) and len(text_chunk.text) > 0:
                result.append(offset)
        return result

    def get_text_chunk_at(self, char_offset):
        """
        Get the (last) text chunk at a given char offset.

        Args:
            char_offset: The char offset of the TextChunk

        Returns:
            The text chunk
        """
        i = 0
        for i in range(len(self.text_chunks) - 1):
            if self.offsets[i + 1]["start"] > char_offset:
                break
        return self.text_chunks[i]

    def get_common_tags(self):
        """
        Returns the list of SAX tags that apply to the whole text block.

        Returns:
            List of common SAX tags
        """
        if len(self.text_chunks) == 0:
            return []

        common_tags = self.text_chunks[0].tags[:]
        for text_chunk in self.text_chunks:
            tags = text_chunk.tags
            if len(tags) < len(common_tags):
                common_tags = common_tags[: len(tags)]

            for j in range(len(common_tags)):
                if common_tags[j]["name"] != tags[j]["name"]:
                    common_tags = common_tags[:j]
                    break

        return common_tags

    def translate_tags(self, target_text, range_mappings):
        """
        Create a new TextBlock, applying our annotations to a translation.

        Args:
            target_text: Translated plain text
            range_mappings: Array of source-target range index mappings

        Returns:
            Translated textblock with tags applied
        """
        # map of { offset: x, text_chunks: [...] }
        empty_text_chunks = {}
        empty_text_chunk_offsets = []
        # list of { start: x, length: x, text_chunk: x }
        text_chunks = []

        def push_empty_text_chunks(offset, chunks):
            for chunk in chunks:
                text_chunks.append({"start": offset, "length": 0, "text_chunk": chunk})

        # Create map of empty text chunks, by offset
        for i, text_chunk in enumerate(self.text_chunks):
            offset = self.offsets[i]["start"]
            if len(text_chunk.text) > 0:
                continue
            if offset not in empty_text_chunks:
                empty_text_chunks[offset] = []
            empty_text_chunks[offset].append(text_chunk)

        empty_text_chunk_offsets = sorted(empty_text_chunks.keys())

        for range_mapping in range_mappings:
            # Copy tags from source text start offset
            source_range_end = range_mapping["source"]["start"] + range_mapping["source"]["length"]
            target_range_end = range_mapping["target"]["start"] + range_mapping["target"]["length"]
            source_text_chunk = self.get_text_chunk_at(range_mapping["source"]["start"])
            text = target_text[range_mapping["target"]["start"] : target_range_end]
            text_chunks.append(
                {
                    "start": range_mapping["target"]["start"],
                    "length": range_mapping["target"]["length"],
                    "text_chunk": TextChunk(text, source_text_chunk.tags, source_text_chunk.inline_content),
                }
            )

            # Empty source text chunks will not be represented in the target plaintext
            j = 0
            while j < len(empty_text_chunk_offsets):
                offset = empty_text_chunk_offsets[j]
                # Check whether chunk is in range
                if offset < range_mapping["source"]["start"] or offset > source_range_end:
                    j += 1
                    continue
                # Push chunk into target text at the current point
                push_empty_text_chunks(target_range_end, empty_text_chunks[offset])
                # Remove chunk from remaining list
                del empty_text_chunks[offset]
                empty_text_chunk_offsets.pop(j)

        # Sort by start position
        text_chunks.sort(key=lambda x: x["start"])

        # Fill in any text_chunk gaps using text with common_tags
        pos = 0
        common_tags = self.get_common_tags()
        i = 0
        while i < len(text_chunks):
            text_chunk = text_chunks[i]
            if text_chunk["start"] < pos:
                raise Exception(f"Overlapping chunks at pos={pos}, text_chunks={i} start={text_chunk['start']}")
            elif text_chunk["start"] > pos:
                # Unmapped chunk: insert plaintext and adjust offset
                text_chunks.insert(
                    i,
                    {
                        "start": pos,
                        "length": text_chunk["start"] - pos,
                        "text_chunk": TextChunk(target_text[pos : text_chunk["start"]], common_tags),
                    },
                )
                i += 1
            pos = text_chunk["start"] + text_chunk["length"]
            i += 1

        # Get trailing text and trailing whitespace
        tail = target_text[pos:]
        import re

        tail_space_match = re.search(r"\s*$", tail)
        tail_space = tail_space_match.group(0) if tail_space_match else ""
        if tail_space:
            tail = tail[: -len(tail_space)]

        if tail:
            # Append tail as text with common_tags
            text_chunks.append({"start": pos, "length": len(tail), "text_chunk": TextChunk(tail, common_tags)})
            pos += len(tail)

        # Copy any remaining text_chunks that have no text
        for offset in empty_text_chunk_offsets:
            push_empty_text_chunks(pos, empty_text_chunks[offset])

        if tail_space:
            # Append tail_space as text with common_tags
            text_chunks.append(
                {"start": pos, "length": len(tail_space), "text_chunk": TextChunk(tail_space, common_tags)}
            )

        return TextBlock([x["text_chunk"] for x in text_chunks])

    def get_plain_text(self):
        """
        Return plain text representation of the text block.

        Returns:
            Plain text representation
        """
        return "".join(chunk.text for chunk in self.text_chunks)

    def get_html(self):
        """
        Return HTML representation of the text block.

        Returns:
            HTML representation
        """
        html = []
        # Start with no tags open
        old_tags = []

        for text_chunk in self.text_chunks:
            # Compare tag stacks; render close tags and open tags as necessary
            # Find the highest offset up to which the tags match
            match_top = -1
            min_length = min(len(old_tags), len(text_chunk.tags))
            for j in range(min_length):
                if old_tags[j] is text_chunk.tags[j]:
                    match_top = j
                else:
                    break

            for j in range(len(old_tags) - 1, match_top, -1):
                html.append(Utils.get_close_tag_html(old_tags[j]))

            for j in range(match_top + 1, len(text_chunk.tags)):
                html.append(Utils.get_open_tag_html(text_chunk.tags[j]))

            old_tags = text_chunk.tags

            # Now add text and inline content
            html.append(Utils.esc(text_chunk.text))
            if text_chunk.inline_content:
                if hasattr(text_chunk.inline_content, "get_html"):
                    # a sub-doc
                    html.append(text_chunk.inline_content.get_html())
                else:
                    # an empty inline tag
                    html.append(Utils.get_open_tag_html(text_chunk.inline_content))
                    html.append(Utils.get_close_tag_html(text_chunk.inline_content))

        # Finally, close any remaining tags
        for j in range(len(old_tags) - 1, -1, -1):
            html.append(Utils.get_close_tag_html(old_tags[j]))

        return "".join(html)

    def get_root_item(self):
        """
        Get a root item in the textblock.

        Returns:
            Root item or None
        """
        for text_chunk in self.text_chunks:
            if len(text_chunk.tags) == 0 and text_chunk.text and re.search(r"[^\s]", text_chunk.text):
                # No tags in this textchunk. See if there is non whitespace text
                return None

            for tag in text_chunk.tags:
                if tag:
                    return tag

            if text_chunk.inline_content:
                inline_doc = text_chunk.inline_content
                # Presence of get_root_item confirms that inline_doc is a Doc instance
                if hasattr(inline_doc, "get_root_item"):
                    root_item = inline_doc.get_root_item()
                    return root_item or None
                else:
                    return inline_doc

        return None

    def get_tag_for_id(self):
        """
        Get a tag that can represent this textblock.

        Returns:
            Tag object
        """
        return self.get_root_item()

    def segment(self, get_boundaries, get_next_id):
        """
        Segment the text block into sentences.

        Args:
            get_boundaries: Function taking plaintext, returning offset array
            get_next_id: Function taking 'segment'|'link', returning next ID

        Returns:
            Segmented version, with added span tags
        """
        # Setup: current_text_chunks for current segment, and all_text_chunks for all segments
        all_text_chunks = []
        current_text_chunks = []

        def flush_chunks():
            if len(current_text_chunks) == 0:
                return

            modified_text_chunks = Utils.add_common_tag(
                current_text_chunks,
                {"name": "span", "attributes": {"class": "cx-segment", "data-segmentid": get_next_id("segment")}},
            )
            Utils.set_link_ids_in_place(modified_text_chunks, get_next_id)
            all_text_chunks.extend(modified_text_chunks)
            current_text_chunks.clear()

        root_item = self.get_root_item()
        if root_item and Utils.is_transclusion(root_item):
            # Avoid segmenting inside transclusions
            return self

        # for each chunk, split at any boundaries that occur inside the chunk
        groups = Utils.get_chunk_boundary_groups(
            get_boundaries(self.get_plain_text()), self.text_chunks, lambda text_chunk: len(text_chunk.text)
        )

        offset = 0
        for group in groups:
            text_chunk = group["chunk"]
            boundaries = group["boundaries"]

            for boundary in boundaries:
                rel_offset = boundary - offset
                if rel_offset == 0:
                    flush_chunks()
                else:
                    left_part = TextChunk(text_chunk.text[:rel_offset], text_chunk.tags[:])
                    right_part = TextChunk(text_chunk.text[rel_offset:], text_chunk.tags[:], text_chunk.inline_content)
                    current_text_chunks.append(left_part)
                    offset += rel_offset
                    flush_chunks()
                    text_chunk = right_part

            # Even if the text_chunk is zero-width, it may have references
            current_text_chunks.append(text_chunk)
            offset += len(text_chunk.text)

        flush_chunks()
        return TextBlock(all_text_chunks)

    def set_link_ids(self, get_next_id):
        """
        Set the link Ids for the links in all the textchunks in the textblock instance.

        Args:
            get_next_id: Function taking 'segment'|'link', returning next ID

        Returns:
            Self with link IDs set
        """
        Utils.set_link_ids_in_place(self.text_chunks, get_next_id)
        return self

    def dump_xml_array(self, pad):
        """
        Dump an XML Array version of the linear representation, for debugging.

        Args:
            pad: Whitespace to indent XML elements

        Returns:
            Array that will concatenate to an XML string representation
        """
        dump = []
        for chunk in self.text_chunks:
            tags_dump = Utils.dump_tags(chunk.tags)
            tags_attr = f' tags="{tags_dump}"' if tags_dump else ""

            if chunk.text:
                dump.append(
                    f"{pad}<cxtextchunk{tags_attr}>" + Utils.esc(chunk.text).replace("\n", "&#10;") + "</cxtextchunk>"
                )

            if chunk.inline_content:
                dump.append(f"{pad}<cxinlineelement{tags_attr}>")
                if hasattr(chunk.inline_content, "dump_xml_array"):
                    # sub-doc: concatenate
                    dump.extend(chunk.inline_content.dump_xml_array(pad + "  "))
                else:
                    dump.append(f'{pad}  <{chunk.inline_content["name"]}/>')
                dump.append(f"{pad}</cxinlineelement>")

        return dump

File: js/lib/lineardoc/text_block.js
import TextChunk from './text_chunk.js';
import { add_common_tag, dump_tags, esc, get_chunk_boundary_groups, get_close_tag_html, get_open_tag_html, is_transclusion, is_transclusion_fragment, set_link_ids_in_place } from './utils.js';
import { getProp } from './util.js';

/**
 * A block of annotated inline text
 *
 * @class
 */
class TextBlock {
	/**
	 * @constructor
	 *
	 * @param {string} text_chunks Annotated inline text
	 * @param {boolean} can_segment This is a block which can be segmented
	 */
	constructor(text_chunks, can_segment) {
		this.text_chunks = text_chunks;
		this.can_segment = can_segment;
		this.offsets = [];
		let cursor = 0;
		for (let i = 0, len = this.text_chunks.length; i < len; i++) {
			this.offsets[i] = {
				start: cursor,
				length: this.text_chunks[i].text.length,
				tags: this.text_chunks[i].tags
			};
			cursor += this.offsets[i].length;
		}
	}

	/**
	 * Get the start and length of each non-common annotation
	 *
	 * @return {Object[]}
	 * @return {number} [i].start {number} Position of each text chunk
	 * @return {number} [i].length {number} Length of each text chunk
	 */
	get_tag_offsets() {
		const textBlock = this,
			commonTags = this.get_common_tags();
		return this.offsets.filter((offset, i) => {
			const text_chunk = textBlock.text_chunks[i];
			return text_chunk.tags.length > commonTags.length && text_chunk.text.length > 0;
		});
	}

	/**
	 * Get the (last) text chunk at a given char offset
	 *
	 * @method
	 * @param {number} charOffset The char offset of the text_chunk
	 * @return {TextChunk} The text chunk
	 */
	get_text_chunk_at(charOffset) {
		let i, len;
		// TODO: bisecting instead of linear search
		for (i = 0, len = this.text_chunks.length - 1; i < len; i++) {
			if (this.offsets[i + 1].start > charOffset) {
				break;
			}
		}
		return this.text_chunks[i];
	}

	/**
	 * Returns the list of SAX tags that apply to the whole text block
	 *
	 * @return {Object[]} List of common SAX tags
	 */
	get_common_tags() {
		if (this.text_chunks.length === 0) {
			return [];
		}
		const commonTags = this.text_chunks[0].tags.slice();
		for (let i = 0, iLen = this.text_chunks.length; i < iLen; i++) {
			const tags = this.text_chunks[i].tags;
			if (tags.length < commonTags.length) {
				commonTags.splice(tags.length);
			}
			for (let j = 0, jLen = commonTags.length; j < jLen; j++) {
				if (commonTags[j].name !== tags[j].name) {
					// truncate
					commonTags.splice(j);
					break;
				}
			}
		}
		return commonTags;
	}

	/**
	 * Create a new text_block, applying our annotations to a translation
	 *
	 * @method
	 * @param {string} targetText Translated plain text
	 * @param {Object[]} rangeMappings Array of source-target range index mappings
	 * @return {text_block} Translated textblock with tags applied
	 */
	translate_tags(targetText, rangeMappings) {
		// map of { offset: x, text_chunks: [...] }
		const emptyTextChunks = {};
		const emptyTextChunkOffsets = [];
		// list of { start: x, length: x, text_chunk: x }
		const text_chunks = [];

		function pushEmptyTextChunks(offset, chunks) {
			for (let c = 0, cLen = chunks.length; c < cLen; c++) {
				text_chunks.push({
					start: offset,
					length: 0,
					text_chunk: chunks[c]
				});
			}
		}

		// Create map of empty text chunks, by offset
		for (let i = 0, iLen = this.text_chunks.length; i < iLen; i++) {
			const text_chunk = this.text_chunks[i];
			const offset = this.offsets[i].start;
			if (text_chunk.text.length > 0) {
				continue;
			}
			if (!emptyTextChunks[offset]) {
				emptyTextChunks[offset] = [];
			}
			emptyTextChunks[offset].push(text_chunk);
		}
		for (const offset in emptyTextChunks) {
			emptyTextChunkOffsets.push(offset);
		}
		emptyTextChunkOffsets.sort((a, b) => a - b);

		for (let i = 0, iLen = rangeMappings.length; i < iLen; i++) {
			// Copy tags from source text start offset
			const rangeMapping = rangeMappings[i];
			const sourceRangeEnd = rangeMapping.source.start + rangeMapping.source.length;
			const targetRangeEnd = rangeMapping.target.start + rangeMapping.target.length;
			const sourceTextChunk = this.get_text_chunk_at(rangeMapping.source.start);
			const text = targetText.slice(rangeMapping.target.start, rangeMapping.target.start + rangeMapping.target.length);
			text_chunks.push({
				start: rangeMapping.target.start,
				length: rangeMapping.target.length,
				text_chunk: new TextChunk(
					text, sourceTextChunk.tags, sourceTextChunk.inline_content
				)
			});

			// Empty source text chunks will not be represented in the target plaintext
			// (because they have no plaintext representation). Therefore we must clone each
			// one manually into the target rich text.

			// Iterate through all remaining emptyTextChunks
			for (let j = 0; j < emptyTextChunkOffsets.length; j++) {
				const offset = emptyTextChunkOffsets[j];
				// Check whether chunk is in range
				if (offset < rangeMapping.source.start || offset > sourceRangeEnd) {
					continue;
				}
				// Push chunk into target text at the current point
				pushEmptyTextChunks(targetRangeEnd, emptyTextChunks[offset]);
				// Remove chunk from remaining list
				delete emptyTextChunks[offset];
				emptyTextChunkOffsets.splice(j, 1);
				// Decrement pointer to match removal
				j--;
			}
		}
		// Sort by start position
		text_chunks.sort((textChunk1, textChunk2) => textChunk1.start - textChunk2.start);
		// Fill in any text_chunk gaps using text with commonTags
		let pos = 0;
		const commonTags = this.get_common_tags();
		for (let i = 0, iLen = text_chunks.length; i < iLen; i++) {
			const text_chunk = text_chunks[i];
			if (text_chunk.start < pos) {
				throw new Error('Overlappping chunks at pos=' + pos + ', text_chunks=' + i + ' start=' + text_chunk.start);
			} else if (text_chunk.start > pos) {
				// Unmapped chunk: insert plaintext and adjust offset
				text_chunks.splice(i, 0, {
					start: pos,
					length: text_chunk.start - pos,
					text_chunk: new TextChunk(
						targetText.slice(pos, text_chunk.start), commonTags
					)
				});
				i++;
				iLen++;
			}
			pos = text_chunk.start + text_chunk.length;
		}

		// Get trailing text and trailing whitespace
		let tail = targetText.slice(pos);
		const tailSpace = tail.match(/\s*$/)[0];
		if (tailSpace) {
			tail = tail.slice(0, tail.length - tailSpace.length);
		}

		if (tail) {
			// Append tail as text with commonTags
			text_chunks.push({
				start: pos,
				length: tail.length,
				text_chunk: new TextChunk(tail, commonTags)
			});
			pos += tail.length;
		}

		// Copy any remaining text_chunks that have no text
		for (let i = 0, iLen = emptyTextChunkOffsets.length; i < iLen; i++) {
			const offset = emptyTextChunkOffsets[i];
			pushEmptyTextChunks(pos, emptyTextChunks[offset]);
		}
		if (tailSpace) {
			// Append tailSpace as text with commonTags
			text_chunks.push({
				start: pos,
				length: tailSpace.length,
				text_chunk: new TextChunk(tailSpace, commonTags)
			});
			pos += tail.length;
		}
		return new TextBlock(text_chunks.map((x) => x.text_chunk));
	}

	/**
	 * Return plain text representation of the text block
	 *
	 * @return {string} Plain text representation
	 */
	get_plain_text() {
		const text = [];
		for (let i = 0, len = this.text_chunks.length; i < len; i++) {
			text.push(this.text_chunks[i].text);
		}
		return text.join('');
	}

	/**
	 * Return HTML representation of the text block
	 *
	 * @return {string} Plain text representation
	 */
	get_html() {
		const html = [];
		// Start with no tags open
		let oldTags = [];
		for (let i = 0, iLen = this.text_chunks.length; i < iLen; i++) {
			const text_chunk = this.text_chunks[i];

			// Compare tag stacks; render close tags and open tags as necessary
			// Find the highest offset up to which the tags match on
			let matchTop = -1;
			const minLength = Math.min(oldTags.length, text_chunk.tags.length);
			for (let j = 0, jLen = minLength; j < jLen; j++) {
				if (oldTags[j] === text_chunk.tags[j]) {
					matchTop = j;
				} else {
					break;
				}
			}
			for (let j = oldTags.length - 1; j > matchTop; j--) {
				html.push(get_close_tag_html(oldTags[j]));
			}
			for (let j = matchTop + 1, jLen = text_chunk.tags.length; j < jLen; j++) {
				html.push(get_open_tag_html(text_chunk.tags[j]));
			}
			oldTags = text_chunk.tags;

			// Now add text and inline content
			html.push(esc(text_chunk.text));
			if (text_chunk.inline_content) {
				if (text_chunk.inline_content.get_html) {
					// a sub-doc
					html.push(text_chunk.inline_content.get_html());
				} else {
					// an empty inline tag
					html.push(get_open_tag_html(text_chunk.inline_content));
					html.push(get_close_tag_html(text_chunk.inline_content));
				}
			}
		}
		// Finally, close any remaining tags
		for (let j = oldTags.length - 1; j >= 0; j--) {
			html.push(get_close_tag_html(oldTags[j]));
		}
		return html.join('');
	}

	/**
	 * Get a root item in the textblock
	 *
	 * @return {Object}
	 */
	getRootItem() {
		for (let i = 0, iLen = this.text_chunks.length; i < iLen; i++) {
			const text_chunk = this.text_chunks[i];

			if (text_chunk.tags.length === 0 && text_chunk.text && text_chunk.text.match(/[^\s]/)) {
				// No tags in this textchunk. See if there is non whitespace text
				return null;
			}

			for (let j = 0, jLen = text_chunk.tags.length; j < jLen; j++) {
				if (text_chunk.tags[j]) {
					return text_chunk.tags[j];
				}
			}
			if (text_chunk.inline_content) {
				const inlineDoc = text_chunk.inline_content;
				// Presence of inlineDoc.getRootItem confirms that inlineDoc is a Doc instance.
				if (inlineDoc && inlineDoc.getRootItem) {
					const rootItem = inlineDoc.getRootItem();
					return rootItem || null;
				} else {
					return inlineDoc;
				}
			}
		}
		return null;
	}

	/**
	 * Get a tag that can represent this textblock.
	 * Textblock can have multiple tags. The first tag is returned.
	 * If there is no tags, but inline_content present, then that is returned.
	 * This is used to extract a unique identifier for the textblock at
	 * Doc#wrapSections.
	 *
	 * @return {Object}
	 */
	getTagForId() {
		return this.getRootItem();
	}

	/**
	 * Segment the text block into sentences
	 *
	 * @method
	 * @param {Function} getBoundaries Function taking plaintext, returning offset array
	 * @param {Function} getNextId Function taking 'segment'|'link', returning next ID
	 * @return {text_block} Segmented version, with added span tags
	 */
	segment(getBoundaries, getNextId) {
		// Setup: currentTextChunks for current segment, and allTextChunks for all segments
		const allTextChunks = [];
		let currentTextChunks = [];
		function flushChunks() {
			if (currentTextChunks.length === 0) {
				return;
			}
			const modifiedTextChunks = add_common_tag(currentTextChunks, {
				name: 'span',
				attributes: {
					class: 'cx-segment',
					'data-segmentid': getNextId('segment')
				}
			});
			set_link_ids_in_place(modifiedTextChunks, getNextId);
			allTextChunks.push.apply(allTextChunks, modifiedTextChunks);
			currentTextChunks = [];
		}

		const rootItem = this.getRootItem();
		if (rootItem && is_transclusion(rootItem)) {
			// Avoid segmenting inside transclusions.
			return this;
		}

		// for each chunk, split at any boundaries that occur inside the chunk
		const groups = get_chunk_boundary_groups(
			getBoundaries(this.get_plain_text()),
			this.text_chunks,
			(text_chunk) => text_chunk.text.length
		);
		let offset = 0;
		for (let i = 0, iLen = groups.length; i < iLen; i++) {
			const group = groups[i];
			let text_chunk = group.chunk;
			const boundaries = group.boundaries;
			for (let j = 0, jLen = boundaries.length; j < jLen; j++) {
				const relOffset = boundaries[j] - offset;
				if (relOffset === 0) {
					flushChunks();
				} else {
					const leftPart = new TextChunk(
						text_chunk.text.slice(0, relOffset), text_chunk.tags.slice()
					);
					const rightPart = new TextChunk(
						text_chunk.text.slice(relOffset),
						text_chunk.tags.slice(),
						text_chunk.inline_content
					);
					currentTextChunks.push(leftPart);
					offset += relOffset;
					flushChunks();
					text_chunk = rightPart;
				}
			}
			// Even if the text_chunk is zero-width, it may have references
			currentTextChunks.push(text_chunk);
			offset += text_chunk.text.length;
		}
		flushChunks();
		return new TextBlock(allTextChunks);
	}

	/**
	 * Set the link Ids for the links in all the textchunks in the textblock instance.
	 *
	 * @param {Function} getNextId Function taking 'segment'|'link', returning next ID
	 * @return {text_block} Segmented version, with added span tags
	 */
	setLinkIds(getNextId) {
		set_link_ids_in_place(this.text_chunks, getNextId);
		return this;
	}

	/**
	 * Adapt a text block.
	 *
	 * @param {Function} getAdapter A function that returns an adapter for the given node item
	 * @return {Promise} Promise that resolves the adapted text_block instance
	 */
	adapt(getAdapter) {
		const textChunkPromises = [];

		// Note that we are not using `await` for the better readable code here. `await` will pause
		// the execution till the `async` call is resolved. For us, while looping over these text
		// chunks and tags, this will create a problem. Adaptations often perform asynchrounous API
		// calls to a MediaWiki instance. If we do API calls for each and every item like a link
		// title, it is inefficient. The API accepts a batched list of titles. We do have a batched
		// API mechanism in cxserver, but that works by debouncing the incoming requests with a
		// timeout. Pausing execution here will cause that debounce handler to be called.
		// So we avoid that pausing by just using an array of promises.
		this.text_chunks.forEach((chunk) => {
			const tagPromises = [],
				tags = chunk.tags;
			tags.forEach((tag) => {
				const dataCX = getProp(['attributes', 'data-cx'], tag);
				if (dataCX && Object.keys(JSON.parse(dataCX)).length) {
					// Already adapted
					return;
				}
				const adapter = getAdapter(tag);
				if (adapter && !is_transclusion_fragment(tag)) {
					// This loop get executed for open and close for the tag.
					// Use data-cx to mark this tag processed. The actual adaptation
					// process below will update this value.
					tag.attributes['data-cx'] = JSON.stringify({ adapted: false });
					tagPromises.push(adapter.adapt());
				}
			});
			textChunkPromises.push(Promise.all(tagPromises));
			let adaptPromise;
			if (chunk.inline_content) {
				if (chunk.inline_content.adapt) {
					// Inline content is a sub document.
					adaptPromise = chunk.inline_content.adapt(getAdapter);
				} else {
					// Inline content is inline empty tag. Examples are link, meta etc.
					const adapter = getAdapter(chunk.inline_content);
					if (adapter && !is_transclusion_fragment(chunk.inline_content)) {
						adaptPromise = adapter.adapt();
					}
				}

				if (adaptPromise) {
					textChunkPromises.push(((chk) => adaptPromise
						.then((adaptedInlineContent) => {
							chk.inline_content = adaptedInlineContent;
						}))(chunk));
				}
			}
		});

		return Promise.all(textChunkPromises).then(() => this);
	}

	/**
	 * Dump an XML Array version of the linear representation, for debugging
	 *
	 * @method
	 * @param {string} pad Whitespace to indent XML elements
	 * @return {string[]} Array that will concatenate to an XML string representation
	 */
	dump_xml_array(pad) {
		const dump = [];
		for (let i = 0, len = this.text_chunks.length; i < len; i++) {
			const chunk = this.text_chunks[i];
			const tagsDump = dump_tags(chunk.tags);
			const tagsAttr = tagsDump ? ' tags="' + tagsDump + '"' : '';
			if (chunk.text) {
				dump.push(pad + '<cxtextchunk' + tagsAttr + '>' +
					esc(chunk.text).replace(/\n/g, '&#10;') +
					'</cxtextchunk>');
			}
			if (chunk.inline_content) {
				dump.push(pad + '<cxinlineelement' + tagsAttr + '>');
				if (chunk.inline_content.dump_xml_array) {
					// sub-doc: concatenate
					dump.push.apply(dump, chunk.inline_content.dump_xml_array(pad + '  '));
				} else {
					dump.push(pad + '  <' + chunk.inline_content.name + '/>');
				}
				dump.push(pad + '</cxinlineelement>');
			}
		}
		return dump;
	}
}

export default TextBlock;
